content/learning-paths/smartphones-and-mobile/android_webgpu_dawn/6-render-a-simple-3D-object-part-2.md

```diff
@@ -0,0 +1,124 @@
+---
+title: Render a simple 3D object - part 2
+weight: 7
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+## 3D meshes
+
+Once Render Pipeline is created, using WebGPU APIs to create and render a 3D mesh is very similar to other graphics APIs. The flow follows:
+
+* Creating Vertex Buffer(s)
+* Creating Index Buffer(s)
+* Creating Uniform Buffer(s)
+* Creating Depth Buffer (Z-Buffer algorithm)
+* Creating Depth Texture and TextureView
+* Creating Depth Stencil
+* Creating Transformation and Projection matrices
+
+All these steps are common in graphics programming and WebGPU offers capability to perform all the operations. It is recommended to go through individual chapters in [3D rendering](https://eliemichel.github.io/LearnWebGPU/basic-3d-rendering/index.html) section.
+
+### Creating and loading 3D objects
+
+Blender is a open-source software widely used to create graphical objects. [Install blender](2-env-setup.md#install-blender) explains how to install blender and export a 3D object in .obj format. In our project we use OBJ files to define 3D meshes. Instead of manually parsing OBJ files, we use the [TinyOBJLoader](https://github.com/tinyobjloader/tinyobjloader) library. The file format is not that complex, but parsing files is not the main point of this learning path, and this library has been intensively tested and has a very small footprint.
+
+{{% notice Note %}}
+Exactly one of your source files must define `TINYOBJLOADER_IMPLEMENTATION` before including it:
+
+```C++
+#define TINYOBJLOADER_IMPLEMENTATION // add this to exactly 1 of your C++ files
+#include "tiny_obj_loader.h"
+```
+
+{{% /notice %}}
+
+We have developed a helper function [`loadGeometryFromObj`](https://github.com/varunchariArm/Android_DawnWebGPU/blob/main/app/src/main/cpp/webgpuRenderer.cpp#L475) using the above library.
+
+With all the setup done, we are now ready to render our 3D object
+
+## Rendering using WebGPU
+
+Now that we have done all the setup, we can run a rendering pass and *draw* something onto our *surface*. To encode any commands to be issued to GPU, we need to create a `CommandEncoder`, which is done:
+
+```C++
+wgpu::CommandEncoderDescriptor commandEncoderDesc;
+commandEncoderDesc.label = "Command Encoder";
+wgpu::CommandEncoder encoder = device.createCommandEncoder(commandEncoderDesc);
+```
+
+Next step is to create `RenderPassEncoder` using `encoder.beginRenderPass()` API:
+
+```C++
+wgpu::RenderPassDescriptor renderPassDesc{};
+
+wgpu::RenderPassColorAttachment renderPassColorAttachment{};
+renderPassColorAttachment.view = nextTexture;
+renderPassColorAttachment.resolveTarget = nullptr;
+renderPassColorAttachment.loadOp = wgpu::LoadOp::Clear;
```

Comment:
Load and store ops are an important aspect (and affect bandwidth). Would be good to explain them and when you would want to use which.

Reply:
Added link to colorAttachment for reader to explore